doc: fix the CQL version in the Interfaces table

annastuchlik opened the pull request:
Fix https://github.com/scylladb/scylla-doc-issues/issues/816
Fix https://github.com/scylladb/scylla-docs/issues/1613

This PR fixes the CQL version in the Interfaces page, so that it is the same as in other places across the docs and in sync with the version reported by the ScyllaDB (see https://github.com/scylladb/scylla-doc-issues/issues/816#issuecomment-1173878487).

To make sure the same CQL version is used across the docs, we should use the `|cql-version| `variable rather than hardcode the version number on several pages.
The variable is specified in the conf.py file:
```
# List of substitutions
rst_prolog = """
.. |cql-version| replace:: 3.3.1
"""
```

tzach commented:
@bhalevy is CQL 3.3.1 the right version?

bhalevy commented:
> @bhalevy is CQL 3.3.1 the right version?

I'm not sure exactly.  @psarna / @nyh may know better.
We sure support more advanced features than CQL 3.3.1, like materialized views for example.

Given https://cassandra.apache.org/doc/latest/cassandra/cql/changes.html
I'm not sure if we fully 3.4.0 or not.
Maybe, on top of the baseline version we fully support, we need to list the additional Cassandra CQL features we support, with their respective version.

bhalevy commented on docs/using-scylla/cassandra-compatibility.rst:
```diff
@@ -21,25 +21,25 @@ For all tables below, the following convention is used:
-JMX                             3.11                                           More Below
-------------------------------  ---------------------------------------------  -------------
+------------------------------  -----------------------------------------------------  -------------
+SSTable format (all versions)   3.11(mc / md ), 2.2(la), 2.1.8 (ka)                    md format is supported in Scylla 4.3, 2021.1.0 and later
```
We will also support the `me` format form version 5.1 / 2022.1.
We should make the change in master and backport it to the 5.1 branch.

annastuchlik replied:
@bhalevy I'm going to update the documentation according to the discussion below, but I'll skip this information as there is a separate issue opened to do that: https://github.com/scylladb/scylla-docs/issues/3990
It's going to be documented for the upcoming OSS and Enterprise releases.

bhalevy commented on docs/using-scylla/cassandra-compatibility.rst:
```diff
@@ -21,25 +21,25 @@ For all tables below, the following convention is used:
 Gossip and internal streaming	NC
-------------------------------  ---------------------------------------------  -------------
+------------------------------  -----------------------------------------------------  -------------
 SSL                             NC
```
@elcallio, is our TLS support incompatible with Cassandra's?

nyh replied:
TLS support for CQL, should be the same (I even added cql-pytest tests for that, run "test/cql-pytest/run --ssl" and it runs all the regular tests but connects to CQL with TLS). But if the heading is on some other use of SSL somewhere (maybe internal communication?) then I don't know.

nyh commented:
We started claiming support of CQL 3.3.1 in commit 856d0e40fb5d457f7d31b6b1e650d19569bfdcac, which was six years ago (!), and the current Cassandra version was 2.2.8.

I don't know what this CQL version even means, or if anyone really keeps it up-to-date in Cassandra, let alone Scylla. Note that it is orthogonal to the CQL **protocol** version (which is currently at 5 in Cassandra, 4 at as IIRC) and the Cassandra version. The Cassandra guys keep https://cassandra.apache.org/doc/trunk/cassandra/cql/changes.html but it cannot possibly be comprehensive, right? Is anyone aware of a more comprehensive document on what "3.3.1" means and what do the newer versions mean?

Any any case, I don't think we should mention CQL 3.3.1 at all, I don't see how it is helpful. We added a lot of things that didn't exist in CQL 3.3.1 and we don't want to suggest that we stopped keeping up with Cassandra compatibility six years ago. In my opinion, our documents should claim compatibility with a particular Cassandra version (e.g., Cassandra 3.11 or Cassandra 4.0), not with a "CQL version" - versions which are not well documented nor testable (compatibility with Cassandra we can can, and do in cql-pytest - how do we check compatibility with a CQL version?) and also I don't think anyone is aware of this numbering... Is a person migrating from Cassandra 3.11 or 4.0 using CQL 3.3.1 or something else? How would they even know?

annastuchlik commented:
This comment illustrates well the complexity @nyh is referring to: https://github.com/scylladb/scylla-docs/issues/1613#issuecomment-511309888. Confusion over what we do and don't support in CQL was already present in 2019 and seems to be only greater now.

**Documentation**
I agree we should remove the references to the supported CQL version from the documentation. @tzach are you OK with that?

**ScyllaDB**
ScyllaDB reports CQL 3.3.1 and protocol v4, see https://github.com/scylladb/scylla-doc-issues/issues/816#issuecomment-1173878487. Perhaps that information should be removed from the product?

nyh commented:
Here is what I think (but feel free to suggest otherwise):
1. We shouldn't casually mention things like "cql 3.1.1" on random pages like https://docs.scylladb.com/stable/cql/ddl.html. This number is both meaningless (Cassandra users probably never even heard of this number) and misleading (because we support a lot of things that didn't exist in 3.1.1). The only version number we should mention in each page is the **Scylla** version that this document refers to. And I think we already do.
2. We should have a "compatibility" page which explains in detail exactly (as much as we can...) how Scylla is compatible with Cassandra. We actually have this already - cassandra-compatibility.rst. We should check how accurate it is (e.g, 3.3.1 is mentioned there too).
3. The versions that Scylla "reports" to the user in the protocol is just something that appeases cqlsh and lets it thinks it is talking to a modern-enough Cassandra. I think we should treat these specific numbers as implementation details and we don't need to dwell on them in the documentation.

annastuchlik commented:
> 1. We shouldn't casually mention things like "cql 3.1.1" on random pages like https://docs.scylladb.com/stable/cql/ddl.html. This number is both meaningless (Cassandra users probably never even heard of this number) and misleading (because we support a lot of things that didn't exist in 3.1.1). The only version number we should mention in each page is the **Scylla** version that this document refers to. And I think we already do.

I agree; I'll remove it.

> 2. We should have a "compatibility" page which explains in detail exactly (as much as we can...) how Scylla is compatible with Cassandra. We actually have this already - cassandra-compatibility.rst. We should check how accurate it is (e.g, 3.3.1 is mentioned there too).

I've been trying for some time now to find Scylla-Cassandra compatibility details, and I've got some, but I'm not sure I have everything. I'll update the Cassandra compatibility page to the best of my knowledge and send it to everyone in RnD for verification. The most recent information is from June 2022 and was provided by @psarna:

**About Cassandra:**
_We are certainly not fully compatible with 4.0, there must be multiple things we don't have implemented, among which I have two recent examples:_
https://github.com/scylladb/scylla/issues/9571
https://github.com/scylladb/scylla/issues/10459
_I'm not an expert, but technically we're compatible with 3.11 and partially with 4.0._

**About CQL:**
_We're compatible with CQL v4, with some extensions from v5 (e.g. Duration type). We also have a couple Scylla-specific extensions (BYPASS CACHE, USING TIMEOUT)._


> 3. The versions that Scylla "reports" to the user in the protocol is just something that appeases cqlsh and lets it thinks it is talking to a modern-enough Cassandra. I think we should treat these specific numbers as implementation details and we don't need to dwell on them in the documentation.

It's not documented, but it was confusing to me and made me think we only support 3.3.1. I was just pointing out the discrepancy, it's not about the documentation.

annastuchlik commented:
Info from Tzach: OK

annastuchlik commented:
@tzach @nyh Could you re-review this PR?

I've removed the CQL version from pages other than the Cassandra compatibility page.

On the Cassandra compatibility page, I have:
- updated the description (intro).
- removed the link to the roadmap (there's little value in it now: https://www.scylladb.com/product/#road-map).
- rewritten the Interfaces table to a more flexible format. This allowed me to add more content to the rows.

![image](https://user-images.githubusercontent.com/37244380/187452400-c3c7eec5-87a5-4fd7-babe-33fe54a5a35c.png)

nyh commented on docs/using-scylla/cassandra-compatibility.rst:
```diff
@@ -2,51 +2,63 @@
+     - Comments
+   * - CQL
+     - | Fully compatible with version 3.3.1, with additional features from later CQL versions (for example, :ref:`Duration type <durations>`).
+       | Fully compatible with protocol v4, with additional features from v5.
```
I'm not an expert in this, but how do these "additional features" work? A client that believes it is using protocol v4 can't use any features from v5...

annastuchlik replied:
@psarna Could you answer the question asked by @nyh and suggest how this section can be improved?

annastuchlik replied:
@psarna Could you help?

psarna replied:
They can, and do. The only example I know of is the Duration type (https://github.com/apache/cassandra/blob/387c274a624c2a85036e909af1d734d87ee00541/doc/native_protocol_v5.spec#L860), not present in v4. Users are free to use this type in CQL queries, and drivers tend to be able to recognize it too, because they often support v5, or at least parts of it. For instance, our Rust driver knows about the Duration type, even though it generally targets v4 protocol: https://github.com/scylladb/scylla-rust-driver/blob/4f5a43dd86ba3f99d722f01384d736af3f52e3b9/scylla-cql/src/frame/response/result.rs#L53

annastuchlik replied:
@psarna Thanks! I think we can leave it as is.

nyh commented on docs/using-scylla/cassandra-compatibility.rst:
```diff
@@ -2,51 +2,63 @@
 
 Supported Tools
 ---------------
-(based on Apache Cassandra 3.11 release)
```
Why remove this comment?
We did base the tools we package (nodetool, etc.) on some Cassandra's version, and it means we may not have features from newer Cassandra versions.

annastuchlik replied:
I removed it because:
- The information that ScyllaDB is compatible with Cassandra 3.11 (+extra stuff)  is already mentioned in the introduction, and I'd like to avoid duplication. 
- Every time we add a new tool, it is added to the list of tools on that page (technically, it's a different page included on the compatibility page), but I'm not sure they are all based on 3.11. If they aren't, such a disclaimer may be misleading.
- Even if it is not misleading, I don't see any value in adding the information here.

If you still think the information here may benefit the user, please let me know, and I'll restore it.

annastuchlik replied:
@nyh What do you think?

nyh replied:
The thing is, the "tools" (nodetool, cqlsh, etc.) are separate programs. It's a separate question of what Cassandra version Scylla itself is compatible with - and what Cassandra version its separate tools are based on.

I think - but again I'm not the expert here - I think @avelanarius would know better - that we really did take a specific version of these tools and made only minor improvements to them - unlike Scylla itself which is completely rewritten code maintaining compatibility.

@avelanarius 's commit 801f7848437a926169f263987970c94b5eeb87cf titled "Sync with Cassandra 3.11.13" suggests that, yes, our tools are based specifically on Cassandra 3.11.13. I think it's worth mentioning this somewhere in the documentation.

annastuchlik replied:
@nyh I see, thanks for the explanation. I've restored the Cassandra version in that section.

annastuchlik commented:
@bhalevy I've added the information about the ME format to the table.

bhalevy commented on docs/using-scylla/cassandra-compatibility.rst:
```diff
@@ -36,8 +36,11 @@ Interfaces
-     - 3.11(mc / md ), 2.2(la), 2.1.8 (ka)
-     - ``md`` format is supported in ScyllaDB Open Source 4.3 and later, and ScyllaDB Enterprise 2021.1.0 and later
+     - 3.11(mc / md / me), 2.2(la), 2.1.8 (ka)
+     - | ``me`` - supported in ScyllaDB Open Source 5.1 and ScyllaDB Enterprise 2022.1.0 (and later)
```
this should be 2022.2.0

(2022.1.0 is based on 5.0 that doesn't support the ME format yet)

annastuchlik replied:
@bhalevy Thanks, I've replaced it with 2022.2.0.

bhalevy reviewed (approved):
LGTM regarding the "me" sstable format.

nyh commented:
Looks good to me, although I questioned whether we should remove - or update - the information of which version of Cassandra our tools (nodetool, cqlsh, etc.) are compatible with.
I'll leave the decision to you, and let me know when to commit this PR.

annastuchlik commented:
@psarna resolved our doubts with his [comment](https://github.com/scylladb/scylladb/pull/11320#discussion_r963409492), so I think this PR is ready to be merged.

nyh commented on docs/using-scylla/cassandra-compatibility.rst:
```diff
@@ -63,6 +63,8 @@ Interfaces
 Supported Tools
 ---------------
 
+The tools are based on Apache Cassandra 3.11.
```
According to 801f7848437a926169f263987970c94b5eeb87cf the accurate version is 3.11.13. But I guess that's not very important here, so ok.

annastuchlik commented:
@scylladb/scylla-maint Please backport to 5.1 if possible

annastuchlik commented:
@scylladb/scylla-maint   Could you backport this PR to 5.1? It was pointed out to me that it should be merged as it specifies 5.1-related information. For example, we started supporting the "me" SStable format in 5.1 (now this information is missing from 5.1, it's only available on master).

denesb commented:
> @scylladb/scylla-maint Could you backport this PR to 5.1? It was pointed out to me that it should be merged as it specifies 5.1-related information. For example, we started supporting the "me" SStable format in 5.1 (now this information is missing from 5.1, it's only available on master).

Done.
